### PR TITLE
BUG: User input checks added for Function class

### PR DESCRIPTION
**Pull request by brunosorban**
Enhanced Validation and Default Settings via `_check_user_input` Method for the `Function` class. The features added comprise:

## Pull request type

- [x] Code changes (bugfix, features)
- [ ] Code maintenance (refactoring, formatting, tests)
- [ ] ReadMe, Docs and GitHub updates
- [ ] Other (please describe):

## Checklist

- [ ] Tests for the changes have been added (if needed)
- [x] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [ ] All tests (`pytest --runslow`) have passed locally

## Current behavior

See issue https://github.com/RocketPy-Team/RocketPy/issues/443.

## New behavior
**Type Checking and Conversion for Source Data:**
The method now checks for list and np.ndarray the the data is consistent. It uses NumPy conversion to an array of type np.float64 for input validation. This ensures consistent data type handling and error triggering for non-homogeneous or non-numeric arrays.

**Output Parameter Validation:**
Enhanced checks for the `outputs` parameter have been added. If `outputs` is a string, it's converted into a list. Additionally, the method now raises a `ValueError` if outputs has a dimension greater than 1, enforcing strict dimensionality constraints.

**Input, Interpolation, and Extrapolation Handling for Multidimensional Functions:**
For `sources` with dimensions greater than 2:
The inputs are automatically set to ["Input 1", "Input 2", ...] if they were initially ["Scalar"], and a warning is issued to inform the user of this default setting.
The interpolation method defaults to `"shepard"` for multidimensional `Functions`, with a warning if any other method is specified, indicating limited support for interpolation methods.
The `extrapolation` is set to "natural" if not specified, with a corresponding warning for clarity and default behavior.

**Source Dimensionality Consistency Check:**
The method now verifies the consistency between the dimensions of `source` and the combined length of `inputs` and `outputs`. It raises a `ValueError` if there's a mismatch, ensuring congruence between source data and expected `input`/`output` dimensions.

**Function Source Handling:**
When the source is a function, the method checks and updates the `inputs` and `outputs` to match those of the source function, issuing warnings if there are discrepancies.

These updates aim to make the method more intuitive, user-friendly, and error-resistant, providing more precise feedback and automatic user adjustments. The added validations and default settings should minimize common configuration errors and guide users toward correct usage patterns, particularly in handling multidimensional data and function sources.

**Added Argument Count Check:**
The code now includes a conditional check to compare the length of args (arguments passed to the function) against the expected domain dimensionality (self.__dom_dim__).
If the number of arguments (len(args)) does not match the expected domain dimension (self.__dom_dim__), the function raises a ValueError.
The error message clearly states the discrepancy, indicating both the expected number of arguments and the number provided. For example: "This Function takes 3 arguments, 2 given."

## Breaking change

- [ ] Yes
- [x] No

## Additional information

These checks should avoid common user errors, but more can always be added. Please feel free to give feedback so we can improve it even more.

### CODE REVIEW

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:1]: Please use [isort](https://pycqa.github.io/isort/) to sort the imports in a .py file.

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:22]: Type hints are not supported yet (although I like your approach), so I'm afraid we need to remove the hints before merging.

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2625]: The function doesn't access or set any attribute of the class. So I think we should convert it to a [staticmethod](https://www.geeksforgeeks.org/class-method-vs-static-method-python/). By doing this, we would allow for using the method without initializing the Function class, which may be useful if someone wants to validate the arguments

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2683]: Try to use "is not None" instead of "!= None", this improves readability in Python.

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2644]: I think these two lines should be at a higher level of this method. I suggest moving this to line 2634.

Also, you can do the same thing for the inputs. You're gonna see that this will fix some tests ;)

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2664]: Small comment but still valuable...
Some comment and warning message lines are larger than 88 characters. We'd appreciate if you break it into more lines please.

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2667]:
```suggestion
                if extrapolation is None:
```

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [inline rocketpy/mathutils/function.py:2661]:
```suggestion
                if interpolation not in [None , "shepard"]:
```

This is more direct and readable

**Review by Gui-FernandesBR — CHANGES_REQUESTED** [review]
Hey @brunosorban very nice job starting this one. I think we are going to the correct target.
I have some comments on your additions, please feel free to ask for further details if needed.

a task list before merging:

- [ ] Different tests are not passing. Maybe you were under the false impression since only the black linter. We need to fix that before proceeding (54 failed, 777 passed, 22 skipped, 35 warnings, 69 errors in 49.32s)
- [ ] Add unit tests to cover the new `_check_user_input` method
- [ ] Document the `_check_user_input` (docstrings)
- [ ] Ensure that all input validations are being done inside the new method and there's no redundant comparisons in the rest of the class.

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:1]: Done

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2625]: Updated

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2683]: Done

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2644]: You cannot make `source_dim = source.shape[1]` if the input is a scalar, which currently is a supported inputs from function class

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2664]: Updated

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2667]: Updatex

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:2661]: Updated

Comment (brunosorban) [reply rocketpy/mathutils/function.py:2661]: I applied the requested changes but I don't have the information about the testes you said it didn't pass and the unit tests

File: rocketpy/mathutils/function.py
Comment (brunosorban) [reply rocketpy/mathutils/function.py:22]: I didn't understand which errors you want to change, but feel free remove the ones that are not supported

Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:22]: Type hinting (https://realpython.com/lessons/type-hinting/) is a Python feature that we didn't adopt in RocketPy yet. We had some discussions on top of #444 recently.

Example:
image

I'm removing the hints and annotations from this PR.  Again, I appreciate the intention, but we decided for not accepting hinting in this library yet, this will make maintainers life easier.

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py]: RocketPy adopts pytest suit for maintaining unit tests that cover small parts of the code. The GitHub actions are used to run the pytest module on every Pull Request. It's possible to check that the tests didn't pass: 

https://github.com/RocketPy-Team/RocketPy/pull/451/checks
image

I'm going to fix it as soon as I understand where the bug is.

**Review by Gui-FernandesBR — APPROVED** [review]
Approved by now. More unit tests could be added to improve the code, but I manages to cover at least the very basic functionality of the new function.

There was another problem in the plot2D function, already solved in my last commit.

To be used in the future:
https://pytest-with-eric.com/introduction/pytest-assert-exception/#:~:text=Exceptions%20can%20be%20tested%20in,doesn%27t%20raise%20an%20exception.

**Comment by Gui-FernandesBR** [comment]
As I was the last person to edit the PR and my changes were kinda delicate, I'd like to wait for another review here before merging it into the master branch.

@MateusStano @phmbressan , could one of you review this one? Please don't get biased by my approved review, this one is important for the future releases...

File: rocketpy/mathutils/function.py
Comment (phmbressan) [inline rocketpy/mathutils/function.py:2806]: I don't know if this is considered a higher dimensional output, but this code runs fine:

```py
import numpy as np
from rocketpy import Function

def source(x):
    return np.array([x, x+1, x+2])

f = Function(source)

f.plot()
```

image

I can even differentiate it:

```bash
>>> print(f.differentiate(5))
array([1., 1., 1.])
```

File: rocketpy/mathutils/function.py
Comment (phmbressan) [inline rocketpy/mathutils/function.py:1303]: This line is causing a mistake due the inner workings of shepard interpolation do not expect the data in this format. What I noticed:

1. Shepard iterator check needs to be updated to handle ndarrays. This is easy I may be done with `collections.abc.Iterable` or adding `np.ndarray` to the list of `isinstance`. However this alone does not fix the mistake.
2. The `x` data from shepard expects a stacked format instead of the first argument list. Still working on a solve.

Just so you can compare:

Before

image


After

image

**Review by phmbressan — CHANGES_REQUESTED** [review]
Great checks for input consistency here, but I caught an error in plotting 2D functions due to shepard interpolation that must be fixed before merging.

Requesting changes for now so nobody merges, I will try to fix the Plot2D error meanwhile.

**Comment by Gui-FernandesBR** [comment]
@phmbressan please improve tests so we prevent from getting the same error in the future.

**Comment by phmbressan** [comment]
I have fixed the plotting error that I mentioned on my earlier review and implemented some tests for the values of `shepard` interpolation to prevent this from happening again.

I would rather if someone double checked the changes before merging. Could you review these @Gui-FernandesBR ?

File: rocketpy/mathutils/function.py
Comment (phmbressan) [reply rocketpy/mathutils/function.py]: Solved by [0f76b3e](https://github.com/RocketPy-Team/RocketPy/pull/451/commits/0f76b3e4c4a9c35155a6543e6d59afae6e7ce7b3).

File: rocketpy/mathutils/function.py
Comment (phmbressan) [reply rocketpy/mathutils/function.py]: Applied in [322e332](https://github.com/RocketPy-Team/RocketPy/pull/451/commits/322e33268c8784183c11f66bcf1f35734c10eadb).

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2826]: image

Comment (phmbressan) [reply rocketpy/mathutils/function.py:2826]: Indeed it seems to be a half supported feature, since it cannot be made with array based Functions and there is this odd R1->R1.

The plots work since `matplotlib` supports multiple values on the "y".

I agree moving this PR foward since implementing fully this feature would be out of scope of this PR.

Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2826]: Yes, agreed.

Just realized something here... So the "outputs" argument should always be a list of one single string. Right?

And should we flag a TODO do understand (and maybe remove) this ValueError raise in this part of the code?

**Review by Gui-FernandesBR — APPROVED** [review]
Great job. Overall a good correction

**Comment by Gui-FernandesBR** [comment]
@phmbressan please solve your request changes and merge it into develop. We have other branches coming up tomorrow,

File: rocketpy/mathutils/function.py
Comment (phmbressan) [reply rocketpy/mathutils/function.py:2826]: Indeed, if the Function only supported output dimension is one, then the `outputs` argument should only be a list of a single string.

I do think that we can support multidimensional outputs due the fact `numpy.ndarray` bahaves quite well in this cases, with a behavior basically undistinguished from a number (e.g. `array([1,1]) + array([1,1]) => array([2,2])`).

**Review by phmbressan — APPROVED** [review]

File: rocketpy/mathutils/function.py
Comment (Gui-FernandesBR) [reply rocketpy/mathutils/function.py:2826]: Does that kind of Function make sense to you? A mathematical operator that takes 1 value in the domain and associate it to 3 different values (3 different images?).

Comment (phmbressan) [reply rocketpy/mathutils/function.py:2826]: I believe mathematically it is a well defined operation, but cannot see right now an application for it in the code.

**Pull request merged**
